File: js/src/typed/nodes/CompanderNode.ts
// Copyright (c) 2018 Spotify AB.
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

/* Generated */

import { v4 as uuid } from 'uuid';
import { Command, Node, ContentType } from '../../score';
import ArgMapper from '../../schema/ArgMapper';
import TypedNode from './TypedNode';
import AudioParam from '../params/AudioParam';

/**
 * CompanderNodeDetectionMode.
 */
export enum CompanderNodeDetectionMode {
  MAX = 'max',
  RMS = 'rms'
}
/**
 * CompanderNodeKneeMode.
 */
export enum CompanderNodeKneeMode {
  HARD = 'hard',
  SOFT = 'soft'
}

/**
 * CompanderNodeConfig.
 */
export interface CompanderNodeConfig {
  detectionMode?: CompanderNodeDetectionMode;
  kneeMode?: CompanderNodeKneeMode;
  cutoffs?: number[];
}

/**
 * CompanderNode.
 */
class CompanderNode extends TypedNode {
  /**
   * Unique identifier for the plugin kind this node represents.
   */
  public static PLUGIN_KIND = 'com.nativeformat.plugin.compressor.compander';

  /**
   * Creates a new `string` mapper for the `detectionMode` param.
   */
  private static DETECTION_MODE_CONFIG = ArgMapper.newEnumArg<
    CompanderNodeDetectionMode
  >(
    'detectionMode',
    CompanderNodeDetectionMode.MAX,
    Object.values(CompanderNodeDetectionMode)
  );

  /**
   * Creates a new `string` mapper for the `kneeMode` param.
   */
  private static KNEE_MODE_CONFIG = ArgMapper.newEnumArg<CompanderNodeKneeMode>(
    'kneeMode',
    CompanderNodeKneeMode.HARD,
    Object.values(CompanderNodeKneeMode)
  );

  /**
   * Creates a new `number[]` mapper for the `cutoffs` param.
   */
  private static CUTOFFS_CONFIG = ArgMapper.newFloatListArg('cutoffs', []);

  /**
   * `compressorThresholdDb` param factory.
   */
  private static COMPRESSOR_THRESHOLD_DB_PARAM = AudioParam.newParamMapper(
    'compressorThresholdDb',
    -24
  );

  /**
   * `compressorKneeDb` param factory.
   */
  private static COMPRESSOR_KNEE_DB_PARAM = AudioParam.newParamMapper(
    'compressorKneeDb',
    30
  );

  /**
   * `compressorRatioDb` param factory.
   */
  private static COMPRESSOR_RATIO_DB_PARAM = AudioParam.newParamMapper(
    'compressorRatioDb',
    12
  );

  /**
   * `expanderThresholdDb` param factory.
   */
  private static EXPANDER_THRESHOLD_DB_PARAM = AudioParam.newParamMapper(
    'expanderThresholdDb',
    -24
  );

  /**
   * `expanderKneeDb` param factory.
   */
  private static EXPANDER_KNEE_DB_PARAM = AudioParam.newParamMapper(
    'expanderKneeDb',
    30
  );

  /**
   * `expanderRatioDb` param factory.
   */
  private static EXPANDER_RATIO_DB_PARAM = AudioParam.newParamMapper(
    'expanderRatioDb',
    12
  );

  /**
   * `attack` param factory.
   */
  private static ATTACK_PARAM = AudioParam.newParamMapper('attack', 0.0003);

  /**
   * `release` param factory.
   */
  private static RELEASE_PARAM = AudioParam.newParamMapper('release', 0.25);

  /**
   * Creates a new `CompanderNode` instance.
   */
  constructor(
    id: string,
    public readonly detectionMode: CompanderNodeDetectionMode,
    public readonly kneeMode: CompanderNodeKneeMode,
    public readonly cutoffs: number[],
    public readonly compressorThresholdDb: AudioParam,
    public readonly compressorKneeDb: AudioParam,
    public readonly compressorRatioDb: AudioParam,
    public readonly expanderThresholdDb: AudioParam,
    public readonly expanderKneeDb: AudioParam,
    public readonly expanderRatioDb: AudioParam,
    public readonly attack: AudioParam,
    public readonly release: AudioParam
  ) {
    super(id, CompanderNode.PLUGIN_KIND);
  }

  /**
   * `CompanderNode` factory.
   */
  public static create(
    config: CompanderNodeConfig,
    id: string = uuid()
  ): CompanderNode {
    return new CompanderNode(
      id,
      CompanderNode.DETECTION_MODE_CONFIG.getValueOrThrow(config.detectionMode),
      CompanderNode.KNEE_MODE_CONFIG.getValueOrThrow(config.kneeMode),
      CompanderNode.CUTOFFS_CONFIG.getValueOrThrow(config.cutoffs),
      CompanderNode.COMPRESSOR_THRESHOLD_DB_PARAM.create(),
      CompanderNode.COMPRESSOR_KNEE_DB_PARAM.create(),
      CompanderNode.COMPRESSOR_RATIO_DB_PARAM.create(),
      CompanderNode.EXPANDER_THRESHOLD_DB_PARAM.create(),
      CompanderNode.EXPANDER_KNEE_DB_PARAM.create(),
      CompanderNode.EXPANDER_RATIO_DB_PARAM.create(),
      CompanderNode.ATTACK_PARAM.create(),
      CompanderNode.RELEASE_PARAM.create()
    );
  }

  /**
   * Creates a new `CompanderNode` from a score `Node`.
   */
  public static from(node: Node): CompanderNode {
    if (CompanderNode.PLUGIN_KIND !== node.kind) {
      throw new Error(`Expected plugin kind="${CompanderNode.PLUGIN_KIND}"`);
    }
    return new CompanderNode(
      node.id,
      CompanderNode.DETECTION_MODE_CONFIG.readConfig(node),
      CompanderNode.KNEE_MODE_CONFIG.readConfig(node),
      CompanderNode.CUTOFFS_CONFIG.readConfig(node),
      CompanderNode.COMPRESSOR_THRESHOLD_DB_PARAM.readParam(node),
      CompanderNode.COMPRESSOR_KNEE_DB_PARAM.readParam(node),
      CompanderNode.COMPRESSOR_RATIO_DB_PARAM.readParam(node),
      CompanderNode.EXPANDER_THRESHOLD_DB_PARAM.readParam(node),
      CompanderNode.EXPANDER_KNEE_DB_PARAM.readParam(node),
      CompanderNode.EXPANDER_RATIO_DB_PARAM.readParam(node),
      CompanderNode.ATTACK_PARAM.readParam(node),
      CompanderNode.RELEASE_PARAM.readParam(node)
    );
  }

  /**
   * Configs for this node.
   */
  public getConfig(): CompanderNodeConfig {
    return {
      detectionMode: this.detectionMode,
      kneeMode: this.kneeMode,
      cutoffs: this.cutoffs
    };
  }

  /**
   * Params for this node.
   */
  public getParams(): { [key: string]: Command[] } {
    return {
      compressorThresholdDb: this.compressorThresholdDb.getCommands(),
      compressorKneeDb: this.compressorKneeDb.getCommands(),
      compressorRatioDb: this.compressorRatioDb.getCommands(),
      expanderThresholdDb: this.expanderThresholdDb.getCommands(),
      expanderKneeDb: this.expanderKneeDb.getCommands(),
      expanderRatioDb: this.expanderRatioDb.getCommands(),
      attack: this.attack.getCommands(),
      release: this.release.getCommands()
    };
  }

  /**
   * Inputs for this node.
   */
  public getInputs(): Map<string, ContentType> {
    return new Map([
      ['audio', ContentType.AUDIO],
      ['sidechain', ContentType.AUDIO]
    ]);
  }

  /**
   * Outputs for this node.
   */
  public getOutputs(): Map<string, ContentType> {
    return new Map([['audio', ContentType.AUDIO]]);
  }
}

/**
 * Export.
 */
export default CompanderNode;

File: swift/Sources/Grapher/Nodes/CompanderNode.swift
// Copyright (c) 2018 Spotify AB.
// 
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
// 
//   http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License

/* Generated */

import Foundation

/// A plugin that performs dynamic range compansion (compression + expansion) on an audio stream.
public final class CompanderNode: Node {
    public enum DetectionMode: String, Codable {
        case max = "max"
        case rms = "rms"
    }
    public enum KneeMode: String, Codable {
        case hard = "hard"
        case soft = "soft"
    }

    private enum ConfigKeys: String, CodingKey {
        case detectionMode = "detectionMode"
        case kneeMode = "kneeMode"
        case cutoffs = "cutoffs"
    }

    private enum ParamKeys: String, CodingKey {
        case compressorThresholdDb = "compressorThresholdDb"
        case compressorKneeDb = "compressorKneeDb"
        case compressorRatioDb = "compressorRatioDb"
        case expanderThresholdDb = "expanderThresholdDb"
        case expanderKneeDb = "expanderKneeDb"
        case expanderRatioDb = "expanderRatioDb"
        case attack = "attack"
        case release = "release"
    }

    public override class var inputs: [String: ContentType] {
        return ["audio": .audio, "sidechain": .audio]
    }

    public override class var outputs: [String: ContentType] {
        return ["audio": .audio]
    }

    /// The signal level detection algorithm to use. Possible values are 'max' and 'rms' (root mean square). This configuration is case-sensitive. Any specified configuration not matching 'max' or 'rms' will be automatically set to 'max'.
    public var detectionMode: DetectionMode

    /// The shape of the knee in the compression function. Possible values are hard and soft. This configuration is case-sensitive. Any specified configuration not matching 'hard' or 'soft' will be automatically set to 'hard'.
    public var kneeMode: KneeMode

    /// A list of cutoff frequencies in Hz for multiband compression. If the list is empty, no band splitting will be performed.
    public var cutoffs: [Double]

    /// An audio parameter specifying the threshold (in dB) at which compression will start.
    public var compressorThresholdDb: AudioParam

    /// An audio parameter specifying the range (in dB) above the threshold at which point the compression curve transitions to the ratio portion.
    public var compressorKneeDb: AudioParam

    /// An audio parameter specifying the amount of dB change from input to 1 dB of output in the compressor.
    public var compressorRatioDb: AudioParam

    /// An audio parameter specifying the threshold (in dB) at which expansion will start.
    public var expanderThresholdDb: AudioParam

    /// An audio parameter specifying the range (in dB) above the threshold at which point the expansion curve transitions to the ratio portion.
    public var expanderKneeDb: AudioParam

    /// An audio parameter specifying the amount of dB change from input to 1 dB of output in the expander.
    public var expanderRatioDb: AudioParam

    /// An audio parameter specifying the amount time till the compander starts reducing the gain.
    public var attack: AudioParam

    /// An audio parameter specifying the amount time till the compander starts increasing the gain.
    public var release: AudioParam

    /**
        Designated initializer

        - Parameters:
            - id: Unique identifier of this node.
            - detectionMode: The signal level detection algorithm to use. Possible values are 'max' and 'rms' (root mean square). This configuration is case-sensitive. Any specified configuration not matching 'max' or 'rms' will be automatically set to 'max'.
            - kneeMode: The shape of the knee in the compression function. Possible values are hard and soft. This configuration is case-sensitive. Any specified configuration not matching 'hard' or 'soft' will be automatically set to 'hard'.
            - cutoffs: A list of cutoff frequencies in Hz for multiband compression. If the list is empty, no band splitting will be performed.
            - compressorThresholdDb: An audio parameter specifying the threshold (in dB) at which compression will start.
            - compressorKneeDb: An audio parameter specifying the range (in dB) above the threshold at which point the compression curve transitions to the ratio portion.
            - compressorRatioDb: An audio parameter specifying the amount of dB change from input to 1 dB of output in the compressor.
            - expanderThresholdDb: An audio parameter specifying the threshold (in dB) at which expansion will start.
            - expanderKneeDb: An audio parameter specifying the range (in dB) above the threshold at which point the expansion curve transitions to the ratio portion.
            - expanderRatioDb: An audio parameter specifying the amount of dB change from input to 1 dB of output in the expander.
            - attack: An audio parameter specifying the amount time till the compander starts reducing the gain.
            - release: An audio parameter specifying the amount time till the compander starts increasing the gain.
     */
    public init(id: String = IDGenerator.generateUniqueID(), detectionMode: DetectionMode = .max, kneeMode: KneeMode = .hard, cutoffs: [Double] = [], compressorThresholdDb: AudioParam, compressorKneeDb: AudioParam, compressorRatioDb: AudioParam, expanderThresholdDb: AudioParam, expanderKneeDb: AudioParam, expanderRatioDb: AudioParam, attack: AudioParam, release: AudioParam) {
        self.detectionMode = detectionMode
        self.kneeMode = kneeMode
        self.cutoffs = cutoffs
        self.compressorThresholdDb = compressorThresholdDb
        self.compressorKneeDb = compressorKneeDb
        self.compressorRatioDb = compressorRatioDb
        self.expanderThresholdDb = expanderThresholdDb
        self.expanderKneeDb = expanderKneeDb
        self.expanderRatioDb = expanderRatioDb
        self.attack = attack
        self.release = release

        super.init(id: id, kind: .companderNode)
    }

    // MARK: Codable

    public required init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let config = try container.nestedContainer(keyedBy: ConfigKeys.self, forKey: .config)
        detectionMode = try config.decodeIfPresent(DetectionMode.self, forKey: .detectionMode) ?? .max
        kneeMode = try config.decodeIfPresent(KneeMode.self, forKey: .kneeMode) ?? .hard
        cutoffs = try config.decodeIfPresent([Double].self, forKey: .cutoffs) ?? []
        let params = try container.nestedContainer(keyedBy: ParamKeys.self, forKey: .params)
        compressorThresholdDb = try params.decode(AudioParam.self, forKey: .compressorThresholdDb, initialValue: -24)
        compressorKneeDb = try params.decode(AudioParam.self, forKey: .compressorKneeDb, initialValue: 30)
        compressorRatioDb = try params.decode(AudioParam.self, forKey: .compressorRatioDb, initialValue: 12)
        expanderThresholdDb = try params.decode(AudioParam.self, forKey: .expanderThresholdDb, initialValue: -24)
        expanderKneeDb = try params.decode(AudioParam.self, forKey: .expanderKneeDb, initialValue: 30)
        expanderRatioDb = try params.decode(AudioParam.self, forKey: .expanderRatioDb, initialValue: 12)
        attack = try params.decode(AudioParam.self, forKey: .attack, initialValue: 0.0003)
        release = try params.decode(AudioParam.self, forKey: .release, initialValue: 0.25)

        try super.init(from: decoder)
    }

    public override func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        var config = container.nestedContainer(keyedBy: ConfigKeys.self, forKey: .config)
        try config.encode(detectionMode, forKey: .detectionMode)
        try config.encode(kneeMode, forKey: .kneeMode)
        try config.encode(cutoffs, forKey: .cutoffs)
        var params = container.nestedContainer(keyedBy: ParamKeys.self, forKey: .params)
        try params.encode(compressorThresholdDb, forKey: .compressorThresholdDb)
        try params.encode(compressorKneeDb, forKey: .compressorKneeDb)
        try params.encode(compressorRatioDb, forKey: .compressorRatioDb)
        try params.encode(expanderThresholdDb, forKey: .expanderThresholdDb)
        try params.encode(expanderKneeDb, forKey: .expanderKneeDb)
        try params.encode(expanderRatioDb, forKey: .expanderRatioDb)
        try params.encode(attack, forKey: .attack)
        try params.encode(release, forKey: .release)

        try super.encode(to: encoder)
    }
}
